File: SenLib/Sen1/FileFixes/scripts_scena_dat_us_t0032_dat.cs
using HyoutaUtils;
using System.Collections.Generic;
using System.IO;

namespace SenLib.Sen1.FileFixes {
	public class scripts_scena_dat_us_t0032_dat : FileMod {
		public string GetDescription() {
			return "Fix formatting issues in teahouse festival events.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xe7854ac057166d50ul, 0xd94c340ec39403d2ul, 0x6173ff9fu));
			if (s == null) {
				return null;
			}

			MemoryStream bin = s.CopyToMemoryAndDispose();
			var patcher = new SenScriptPatcher(bin);

			bin.Position = 0x2360;
			bin.WriteUInt24(0x2e2e2e);
			bin.Position = 0xe22a;
			bin.WriteUInt24(0x2e2e2e);
			bin.Position = 0x11fde;
			bin.WriteUInt24(0x2e2e2e);
			bin.Position = 0x137db;
			bin.WriteUInt24(0x2e2e2e);

			// 'no way I can to quit now' -> 'no way I can quit now'
			patcher.RemovePartialCommand(0x1622b, 0xa5, 0x162c2, 0x3);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/t0032.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xe7854ac057166d50ul, 0xd94c340ec39403d2ul, 0x6173ff9fu));
			if (s == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/t0032.dat", s) };
		}
	}
}

File: native/sen1/file_fixes/scripts_scena_dat_us_t0032_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/sen_script_patcher.h"
#include "sha1.h"
#include "util/memwrite.h"

namespace SenLib::Sen1::FileFixes::scripts_scena_dat_us_t0032_dat {
std::string_view GetDescription() {
    return "Fix formatting issues in teahouse festival events.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_us/t0032.dat",
            95586,
            SenPatcher::SHA1FromHexString("e7854ac057166d50d94c340ec39403d26173ff9f"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        using namespace HyoutaUtils::MemWrite;

        SenScriptPatcher patcher(bin);

        WriteArray(&bin[0x2360], {{0x2e, 0x2e, 0x2e}});
        WriteArray(&bin[0xe22a], {{0x2e, 0x2e, 0x2e}});
        WriteArray(&bin[0x11fde], {{0x2e, 0x2e, 0x2e}});
        WriteArray(&bin[0x137db], {{0x2e, 0x2e, 0x2e}});

        // 'no way I can to quit now' -> 'no way I can quit now'
        patcher.RemovePartialCommand(0x1622b, 0xa5, 0x162c2, 0x3);

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen1::FileFixes::scripts_scena_dat_us_t0032_dat
